app/jobs/import_from_vacancy_source_job.rb

```diff
@@ -30,18 +31,27 @@ def import_from_source(source_klass)
           log("Imported vacancy #{vacancy.id}.")
         else
           failure = create_failed_imported_vacancy(vacancy)
-          errors[failure.external_reference] = failure.import_errors
+          @errors[failure.external_reference] = failure.import_errors
         end
       end
     end
-    report_validation_errors(total_count, errors)
-    mark_removed_vacancies_from_source(source_external_references)
   end
 
-  def mark_removed_vacancies_from_source(source_external_references)
+  # Avoids unnecessary and expensive DB updates for vacancies that have not changed since they were imported.
+  #
+  # Before checking for changes it calls a method that sets to 'nil' multiple invalid field values.
+  # This method is called as an AR callback on the vacancy before it is saved.
+  # Calling it here avoids 'changed?' returning true when, after being normalised by the callback and saved,
+  # the DB value wouldn't change.
+  def existing_vacancy_unchanged?(vacancy)
+    vacancy.reset_dependent_fields
```

Comment:
This little f*er called in a `before :save` from a model concern (even more fun to find!) was what caused the `nil` subjects from the feed to turn into `[]` in DB sometimes, plus vacancies from the feed coming with subjects that were never saved  (and the same in other attributes!)